fedora_messaging/twisted/factory.py

```diff
@@ -0,0 +1,156 @@
+        if self._client_ready.called:
+            # Renew the ready deferred, it will callback when the
+            # next connection is ready.
+            self._client_ready = defer.Deferred()
+        protocol.ReconnectingClientFactory.clientConnectionLost(
+            self, connector, reason)
+
+    def clientConnectionFailed(self, connector, reason):
+        log.msg('Connection failed. Reason: {}'.format(reason.value),
+                system=self.name, logLevel=logging.WARNING)
+        protocol.ReconnectingClientFactory.clientConnectionFailed(
+            self, connector, reason)
+
+    def stopTrying(self):
+        protocol.ReconnectingClientFactory.stopTrying(self)
+        if not self._client_ready.called:
+            self._client_ready.errback(pika.exceptions.AMQPConnectionError(
+                "Could not connect, reconnection cancelled."
+            ))
+
+    @defer.inlineCallbacks
+    def stopFactory(self):
+        if self.client:
+            yield self.client.stopProducing()
+        protocol.ReconnectingClientFactory.stopFactory(self)
+
+    @defer.inlineCallbacks
+    def consume(self, message_callback):
+        """Pass incoming messages to the provided callback.
+
+        Args:
+            message_callback (callable): The callable to pass the message to
+                when one arrives.
+        """
+        log.msg('Messages reading setup',
+                system=self.name, logLevel=logging.DEBUG)
+        new_setup = self._message_callback is None
+        self._message_callback = message_callback
+        if self._client_ready.called and new_setup:
+            # If consume() is called after the client is ready (and we did
+            # not setup before), do it now.
+            yield self.client.setupRead(self._message_callback)
+            yield self.client.resumeProducing()
+
+    @defer.inlineCallbacks
+    def publish(self, message, exchange=None):
+        """
+        Publish a :class:`fedora_messaging.message.Message` to an `exchange`_
+        on the message broker.
+
+        Args:
+            message (message.Message): The message to publish.
+            exchange (str): The name of the AMQP exchange to publish to; defaults
+to :ref:`conf-publish-exchange`
+
+        Raises:
+            PublishReturned: If the published message is rejected by the broker.
+            ConnectionException: If a connection error occurs while publishing.
+
+        .. _exchange: https://www.rabbitmq.com/tutorials/amqp-concepts.html#exchanges
+        """
+        yield self._client_ready
+        try:
+            yield self.client.publish(message, exchange)
+        except (
+                pika.exceptions.ConnectionClosed, pika.exceptions.ChannelClosed
+                ) as e:
+            log.msg('Connection lost while publishing, retrying.',
+                    system=self.name, logLevel=logging.WARNING)
+            yield self.publish(message, exchange)
```

Comment:
Will this eventually time out? I guess we do want it to try until it succeeds or the service is shut down.

Reply:
Hmm I don't think it'll time out, it should just wait for the new connection to be ready and publish afterwards (waiting for the `_client_ready` Deferred). Did I understand your question correctly?

Reply:
Yeah, that was what I thought. It may be that someday someone will want to publish something that errbacks after a timeout is reached rather than retrying forever, but that's not something we need to add right now.